### ISSUE TITLE
Potential redundant field in MethodDesc

### ISSUE THREAD

**Issue by fabled** [dotnet/runtime#96474]
PR #85743 added to `class MethodDesc` `SetMethodDescIndex()`, `SetMethodDescIndex()` and made space for the new field `m_methodIndex` at https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/method.hpp#L1637C17-L1637C29.

This looks redundant with the earlier existing (renamed newly in the PR) members `SetChunkIndex()` and `GetChunkIndex()`.

Both seem to hold fundamentally same data. In all three locations these are initialized together. The other one will hold the integer index and the other one the byte offset (with multiplier).

Since the integer index variant is new and needs to be used as such in some locations, that sounds like useful to keep. The `SetChunkIndex` can likely be removed, and callers of `GetMethodDescChunkIndex` can be refacture to use the other field or some new helpers that calculate the data needed.

This would eliminate `m_chunkIndex` from the `class MethodDesc` which needs to be small and seems to be running out of bits. And reduce confusion why there is two similar looking fields with different names but almost same data.

**Comment by jkotas** [comment]
@davidwrighton

**Comment by davidwrighton** [comment]
Yes, the data is irritatingly similar, but not quite the same. What's going on here is that getting the temporary entrypoint for the `MethodDesc` is a common operation during process startup, and before this field existed, we needed to do a search in the list of temporary entrypoints in the `MethodDescChunk` for the `MethodDesc` which was quite slow, as mapping from temporary entry point to `MethodDesc` requires several memory reads. Once we added this field, the operation became O(1), and also with a very small constant factor, as reading a single field is very efficient. Now, overall for program execution, this field is not significant, but for startup, I measured a substantial amount of time improvement from this extra field, and since we have the bits we might as well use them. An alternative approach would have been to adjust the behavior of code around https://github.com/dotnet/runtime/blob/e565285e3347929f7327f96dd58ae88b2de77a28/src/coreclr/vm/methodtablebuilder.cpp#L10850 such that we don't need to do the complex lookup at that point. (That location is the source of the most cost associated with temporary entrypoint lookups.) However, since we had the bits available, it was pretty easy to add this other index, which also provides a small speed up for the other cases where temporary entry points are looked up.

**Comment by fabled** [comment]
Let me rephrase my suggestion.

Is there something that prevents implementing `MethodDesc::GetMethodDescIndex()` as
``` C
  return m_methodIndex * sizeof(MethodDesc) / MethodDesc::ALIGNMENT;
```

After this `m_methodIndex` would never be read and could be eliminated. There would be other immediate simplifications in the code base too.

If there are conditions when `m_methodIndex` and `m_chunkIndex` are not consistent with each other, I would curios to know about it. Thanks.

**Comment by davidwrighton** [comment]
Unfortunately the amount of memory a `MethodDesc` takes up within a `MethodDescChunk` is not the same for each `MethodDesc`. See the implementation of `MethodDesc::SizeOf` for details. Whenever a `MethodDesc` isn't sized as a standard MethodDesc, these `m_methodIndex` and `m_chunkIndex` become less directly correlated.

**Comment by fabled** [comment]
Makes sense now. Seems I was concentrating too much to `MethodDescChunk::CreateChunk` and missed the funky bits in `MethodTableBuilder::AllocAndInitMethodDescChunk`. The latter actually seems somewhat inconsistent as its using `sizeof` for the starting offset at https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/methodtablebuilder.cpp#L7044 instead of the accessor like the other site at https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/method.cpp#L1745.

Anyhow, sorry for the noise, and thanks for the answers.
